KAFKA-2660; Correct cleanableRatio calculation

lindong28 opened the pull request:
@onurkaraman Could you have a look? This is the patch I discussed with you.

lindong28 committed:
KAFKA-2660; Correct cleanableRatio calculation

onurkaraman commented on core/src/main/scala/kafka/log/LogCleaner.scala:
```diff
@@ -261,9 +261,12 @@ class LogCleaner(val config: CleanerConfig,
+ * @param dupBufferLoadFactor The maximum percent full for the deduplication buffer
  * @param throttler The throttler instance to use for limiting I/O rate.
  * @param time The time instance
+ * @param checkDone Check if the cleaning for a partition is aborted. If so, throw an exception
```
I think the description for the checkDone param is somewhat misleading.

Notice that the Cleaner class uses checkDone but doesn't explicitly throw exceptions. It just so happens that the checkDone function passed into the Cleaner from the CleanerThread uses LogCleanerManager.checkCleaningAborted throws a LogCleaningAbortedException.

lindong28 replied:
Thanks for the review! Fixed now.

onurkaraman commented on core/src/main/scala/kafka/log/LogCleaner.scala:
```diff
@@ -673,7 +676,7 @@ private case class CleanerStats(time: Time = SystemTime) {
  */
 private case class LogToClean(topicPartition: TopicAndPartition, log: Log, firstDirtyOffset: Long) extends Ordered[LogToClean] {
-  val cleanBytes = log.logSegments(-1, firstDirtyOffset-1).map(_.size).sum
+  val cleanBytes = log.logSegments(-1, firstDirtyOffset).map(_.size).sum
```
Good catch!

lindong28 committed:
update doc

onurkaraman commented:
LGTM

lindong28 commented:
@jjkoshy This is a very small patch and it has been reviewed by Onur. Do you have time to have a look?

jjkoshy commented:
I agree with the change, but a couple of comments:

I think the way the cleaner works right now, it should be that the dirty offset should always be the start of a segment. So the estimate will be off by at most the size of the last message in the clean section. Can you verify?

Also, it appears from what you described offline that `Log.logSegments` isn't doing what it claims in its doc. I'm not sure that's the case but can you elaborate and if it is an issue consider other usages of the segment range lookup?

Finally, can you add or edit a unit test in `CleanerTest` for this?

lindong28 commented:
@jjkoshy Thanks for your review. Sure, I will provide answer to issues you raised.

lindong28 commented:
@jjkoshy You are right, I have verified that in existing implementation the dirty offset should always be the start of a segment. Therefore the estimate will be off by at most one message size in the clean section.

The documentation of `Log.logSegments(from: Long, to: Long)` would be wrong if either `from` or `to` are is NOT the start of a segment. For example, suppose we have segments with the following offsets: `[1,2,3,4,5], [6,7,8,9], [10,11,12]`, then `Log.logSegments(2, 11)` would only return the segment `[6,7,8]`, even though it should return all 3 segments according to the Doc. In addition, `to` and `from` doesn't seem to always be the start of a segment, e.g. `logSegments(this.recoveryPoint, offset)` in Log.scala.

Commit:
add unit test

lindong28 commented:
@jjkoshy I have added unit test to verify the problem and the fix for LogToClean. I also fixed some minor problem that I found in the process of checking log related code. Thanks for taking time to review the patch.

lindong28 commented:
@guozhangwang Would you have time to review this patch?

lindong28 commented:
@gwenshap Would you have time to review this patch? It may be useful to include in 0.9 release since it fixes a bug in log cleaner.

gwenshap commented:
Super sorry @lindong28, but I just don't know enough about the cleaner to properly review this. 
I hope @guozhangwang, @junrao or @jjkoshy will have time.

junrao commented:
Thanks for the patch. LGTM

lindong28 commented:
@junrao Thanks for the review! @gwenshap Thanks:)
